resource/kind.go

```diff
@@ -0,0 +1,431 @@
+// Group, Version, Kind, Plural, and Scope, as well as allowing for creation of a zero-value version of the Kind as an
+// Object (which may contain default values for fields as decided by the implementation).
+// Additionally, Kind contains the logic for reading wire bytes into an Object instance of the Kind,
+// or serializing an Object instance of the Kind into wire bytes.
+// In this sense, Kind combines schema information and a serializer into one interface used to interact with the kind.
+type Kind interface {
+	KindReadWriter
+	Schema
+}
+
+// NewKindGroup returns a new KindGroup with the provided group and version
+func NewKindGroup(group, version string) *KindGroup {
+	return &KindGroup{
+		group:   group,
+		version: version,
+		kinds:   make([]Kind, 0),
+	}
+}
+
+// KindGroup is a set of Kind objects which share the same Group and Version.
+type KindGroup struct {
+	group   string
+	version string
+	kinds   []Kind
+}
+
+// Group returns the group value shared by all kinds in the KindGroup
+func (k *KindGroup) Group() string {
+	return k.group
+}
+
+// Version returns the version value shared by all kinds in the KindGroup
+func (k *KindGroup) Version() string {
+	return k.version
+}
+
+// Kinds returns the list of kinds registered with the group
+func (k *KindGroup) Kinds() []Kind {
+	return k.kinds
+}
+
+// AddKind adds a Kind to the group, if its Group() and Version() values match the KindGroup's.
+// Otherwise, it returns an error.
+func (k *KindGroup) AddKind(kind Kind) error {
+	if kind == nil {
+		return fmt.Errorf("kind cannot be nil")
+	}
+	if k.kinds == nil {
+		k.kinds = make([]Kind, 0)
+		if k.group == "" {
+			k.group = kind.Group()
+		}
+		if k.version == "" {
+			k.version = kind.Version()
+		}
+	}
+	if k.group != kind.Group() || k.version != kind.Version() {
+		return fmt.Errorf("kind group is restricted to group/version %s/%s, provided kind is %s/%s", k.group, k.version, kind.Group(), kind.Version())
+	}
+	k.kinds = append(k.kinds, kind)
+	return nil
+}
+
+// UntypedKind is a generic implementation of Kind, which will work for any kubernetes kind,
+// with Write consuming any Object2 and producing kubernetes bytes, and Read consuming any valid kubernetes kind bytes
+// and producing an *UntypedKind. The values for all getter methods are set via the corresponding struct fields.
+type UntypedKind struct {
+	// GVK is the group, version, and kind of the UntypedKind, returned by Group(), Version(), and Kind() respectively
+	GVK schema.GroupVersionKind
+	// PluralKind is the plural name of the Kind, returned via Plural()
+	PluralKind string
+	// KindScope is the scope of the Kind, returned by Scope()
+	KindScope SchemaScope
+	// ZeroObject is the *UntypedObject returned by ZeroValue(). This only needs to be non-nil if you do not want to use
+	// an empty *UntypedObject as the ZeroValue() return.
+	ZeroObject *UntypedObject
```

Comment:
Do we have to make it a pointer? In general we should try to avoid pointers – they are prone to null-pointer errors and they always escape to the heap.

Reply:
`*UntypedObject` implements `Object`, `UntypedObject` doesn't. The implementations have to be pointer structs due to the `Set` methods for fields.